jbmc/src/java_bytecode/lambda_synthesis.cpp

```diff
@@ -635,44 +701,40 @@ exprt box_or_unbox_type_if_necessary(
 
   // One is a pointer, the other a primitive -- box or unbox as necessary, and
   // check the types are consistent:
-  if(original_is_pointer)
-  {
-    const irep_idt &boxed_type_id =
-      to_struct_tag_type(original_type.subtype()).get_identifier();
-    const auto *boxed_type_info = get_boxed_type_info_by_name(boxed_type_id);
-    INVARIANT(
-      boxed_type_info != nullptr,
-      "Only boxed primitive types should participate in a pointer vs."
-      " primitive type disagreement");
-
-    symbol_exprt fresh_local = create_and_declare_local(
-      function_id, role + "_unboxed", required_type, symbol_table, code_block);
-    const symbolt &unboxing_function_symbol =
-      symbol_table.lookup_ref(boxed_type_info->unboxing_function_name);
-    code_block.add(code_function_callt{
-      fresh_local, unboxing_function_symbol.symbol_expr(), {expr}});
-
-    return std::move(fresh_local);
-  }
-  else
-  {
-    const auto *primitive_type_info =
-      get_java_primitive_type_info(original_type);
-    INVARIANT(
-      primitive_type_info != nullptr,
-      "A Java non-pointer type involved in a type disagreement should"
-      " be a primitive");
-
-    symbol_exprt fresh_local = create_and_declare_local(
-      function_id, role + "_boxed", required_type, symbol_table, code_block);
-    const symbolt &boxed_type_factory_method =
-      symbol_table.lookup_ref(primitive_type_info->boxed_type_factory_method);
-
-    code_block.add(code_function_callt{
-      fresh_local, boxed_type_factory_method.symbol_expr(), {expr}});
-
-    return std::move(fresh_local);
-  }
+
+  const auto *primitive_type_info = get_java_primitive_type_info(
+    original_is_pointer ? required_type : original_type);
+  INVARIANT(
+    primitive_type_info != nullptr,
+    "A Java non-pointer type involved in a type disagreement should"
```

Comment:
:question: What non-pointer java types are not primitives?

Reply:
Void! Which should have been excluded earlier. I guess this would also fire if you screwed up and fed it a deref'd pointer type, the string_typet @class_identifier, or similar.